designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,213 @@
+| Authors       | Creation Date | Status      | Extra |
+|---------------|---------------|-------------|---|
+| @name | date | Implementeble | - |
+
+# Auto-generate Helm Chart Plugin
+===================
+
+The proposal is to add a new alpha command to auto-generate a helm chart for a given kustomize directory created by kubebuilder.
+
+## Example
+
+By running the following command, the plugin will generate a helm chart from the specific kustomize directory and output it to the directory specified by the `--output` flag.
+
+```shell
+kubebuilder alpha generate-helm-chart --from=<path> --output=<path>
+```
+
+## Open Questions [optional]
+
+TBF
+
+## Summary
+
+TBF
+
+## Motivation
+
+Currently, kubebuilder only supports the distribution of the operator by the kustomize. However, many well-known operators contains the distribution of helm charts:
+
+- [mongodb](https://artifacthub.io/packages/helm/mongodb-helm-charts/community-operator
+)
+- [cert-manager](https://cert-manager.io/v1.6-docs/installation/helm/#1-add-the-jetstack-helm-repository)
+- [prometheus](https://bitnami.com/stack/prometheus-operator/helm)
+
+The proposed command will auto-generate a helm chart from a kustomize directory created by kubebuilder with minimizing the manual effort. The generated helm chart will cover the frequently-used configurations.
+
+Context:
+
+- See the discussion [Is there a good solution/tool to generate a new helm chart from the manifests?](https://github.com/kubernetes-sigs/kubebuilder/discussions/3074)
+
+### Goals
+
+- Help developers to generate a helm chart from a kustomize directory.
+- Generate a helm chart without human interactions once the kustomize directory updated.
+- Reserve the ability to customize the helm chart by the developer.
+
+### Non-Goals
+
+- Generate the helm chart from a kustomize directory not created by kubebuilder.
+- Generate the fine-grained configurations.
+
+## Proposal
+
+To reduce the user adaptation, the command
+leverage the existing kustomize default directory and create the similar layout of the helm chart from it. The generated `values.yaml` is as follows:
```

Comment:
Question?
The values is to allow us edit the values right?
So, should we really add those options?

Reply:
The values file present in a Chart definition is used as default when you pull it as dependency somewhere else. I think it's a good opportunity to set sane defaults for users.

That said, I'm leaning towards thinking we should only generate the values file during scaffolding and then never touch it again automatically (letting the user customize it at will, without `make generate` undoing these customizations).

If we do want to touch it after initial scaffolding, I think we would need some markers to express how these defaults should vary. It would also need to not change **additional** stuff authors add (eg, ConfigMaps if their controller mounts a config file).

Reply:
The values has the manager values 
Those are not changed with make generate manifest

After the first scaffold of the helm chart what mainly would change with `make generate manifest` would be the RBAC and the CRDs, Those we will need to replace. 

But at the same way that we do not change the manager manifests we would not change those ones after the first scaffold.

Reply:
> So, should we really add those options?

I think the answer is yes, these options provide the following chances to adapt the deployment of operator to the kubernetes cluster.

- Deploy the operator on the specific node.
- Adjust the operator's resources limitations.
- Specify the debug version of the operator image.
- ...

Reply:
@dashanji,

But the idea of helmChart is to distribute the Operator itself.
So, my concern here is we allow the helmChart not be sync with the Operator itself by default we can add that in the OpenQuestions.

Reply:
> But the idea of helmChart is to distribute the Operator itself.

Agree, but we can't have no options, as it will be the same as all-in-one YAML. I think the key issue is to find some common and useful options.